### DISCUSSION TITLE
Implicit `break` in switch statements

### DISCUSSION THREAD

**Discussion by scott-parkhill** [dotnet/csharplang#7115] [Language Ideas]
Fall-through is already not allowed in C# if some sort of expression is given, but you still always have to manually give it a `break`. This is cumbersome, annoying, and hurts the readability of switch statements. I personally would use switches far more frequently if I didn't have to specify `break` all the time. 

In the following, I will use this enum as an example:

```
enum Foo
{
    None,
    Bar,
    Baz,
    Boo
}
```

My idea is this:

1. If no fall-throughs occur, and no gotos occur, and no return statements occur within a switch statement, then breaks are taken to be implicit. Otherwise, the compiler will give an error saying that since other flow control has occurred during the switch statement then all fall control must be explicitly specified.

That means that this would compile, and insert `break` in all the appropriate places:

```
switch (myEnum)
{
    case Foo.Bar:
        Method1();
    case Foo.Baz:
        Method2();
    case Foo.Boo:
        Method3();
    default:
        break;
}
```

But this would not compile and instead give the said error message:

```
switch (myEnum)
{
    case Foo.Bar:
        Method1();
        return true;
    case Foo.Baz:
        Method2();
    case Foo.Boo:
        Method3();
    default:
        break;
}
```

And this would also be invalid:

```
switch (myEnum)
{
    case Foo.Bar:
    case Foo.Baz:
        Method1();
    case Foo.Boo:
        Method3();
    default:
        break;
}
```

In these cases, you would have to specify all the appropriate `break`s.

2. Further, `break` may only be used when it is the **only** expression that occurs in a case, and if a `break` occurs after any expression(s) and elsewhere implicit `breaks` are used, this will give a compiler warning or error saying implicit breaking is used but an explicit break has occurred. I'm not sure if warning or error is better here but I lean toward warning since this could occur from habit and have no real consequence.

So, this would compile but generate a warning:

```
switch (myEnum)
{
    case Foo.Bar:
        Method1();
    case Foo.Baz:
        Method2();
        break;
    case Foo.Boo:
        Method3();
    default:
        break;
}
```

3. Just to be extra clear, this:

```
switch (myEnum)
{
    case Foo.Bar:
        Method1();
    case Foo.Baz:
        Method2();
    case Foo.Boo:
        Method3();
    default:
        break;
}
```

would compile into this:

```
switch (myEnum)
{
    case Foo.Bar:
        Method1();
        break;
    case Foo.Baz:
        Method2();
        break;
    case Foo.Boo:
        Method3();
        break;
    default:
        break;
}
```

Thank you for your time and consideration and I look forward to any and all comments and discussion.

Edit: formatting

**Comment by scott-parkhill** [comment]
If this is considered a dangerous change, it could also have an XML boolean in `.csproj`. I.e. `<ImplicitBreaking>enable</ImplicitBreaking>`.

**Reply by HaloFour** [reply]
A project setting that changes the behavior of the language would result in a dialect, which is something the language team is pretty opposed to except in extreme circumstances.

**Reply by scott-parkhill** [reply]
Makes sense, thanks for the clarity.

**Comment by scott-parkhill** [comment]
An open discussion on this is https://github.com/dotnet/csharplang/discussions/5721 but this is more fleshed out here.

**Comment by koszeggy** [comment]
See #2632.

Instead of breaking the original `switch` construct, it would allow this even more compact syntax:
```cs
myEnum switch
{
    Foo.Bar => Method1(),
    Foo.Baz => Method2(),
    Foo.Boo => Method3(),
    _ => WhateverElse()
};
```

**Reply by scott-parkhill** [reply]
Yes, I had seen this proposal and like it, but the OG switch statement is still a good construct and does have a lot of clarity when doing multiple things at once for a case. Yes, you'd be able to just open a scope with the expression syntax, but it's not always as clean imo. I guess, one does not preclude the other.
